### PR TITLE
VIH-6988 Epic private consultations

### PR DESCRIPTION
**Pull request by levibutler**
### JIRA link (if applicable) ###
VIH-6989
VIH-6988
VIH-6521
VIH-6520
VIH-7152
VIH-6990
VIH-7131
VIH-6522

### Change description ###
===BREAKING CHANGES===

**Does this PR introduce a breaking change?** (check one with "x")

```
[X] Yes
[ ] No
```

### CODE REVIEW

File: VideoApi/VideoApi.Services/KinlyPlatformService.cs [hunk 1]
```diff
@@ -176,15 +143,14 @@ public async Task StopPrivateConsultationAsync(Conference conference, RoomType c
             foreach (var participant in participants)
             {
                 await TransferParticipantAsync(conference.Id, participant.Id, consultationRoom,
-                    RoomType.WaitingRoom);
+                    RoomType.WaitingRoom.ToString());
```
Comment (rehmanab) [inline VideoApi/VideoApi.Services/KinlyPlatformService.cs:146]: Null check

File: VideoApi/VideoApi.Services/KinlyPlatformService.cs [hunk 2]
```diff
@@ -176,15 +143,14 @@ public async Task StopPrivateConsultationAsync(Conference conference, RoomType c
...
             }
 
             var endpoints = conference.GetEndpoints()
                 .Where(x => x.State == EndpointState.InConsultation && x.GetCurrentRoom() == consultationRoom);
             foreach (var endpoint in endpoints)
             {
-                await TransferParticipantAsync(conference.Id, endpoint.Id, consultationRoom,
-                    RoomType.WaitingRoom);
+                await TransferParticipantAsync(conference.Id, endpoint.Id, consultationRoom, RoomType.WaitingRoom.ToString());
```
Comment (rehmanab) [inline VideoApi/VideoApi.Services/KinlyPlatformService.cs:153]: Null check

File: VideoApi/VideoApi/Controllers/ConferenceManagementController.cs [hunk 3]
```diff
@@ -145,14 +145,14 @@ public async Task<IActionResult> TransferParticipantAsync(Guid conferenceId, Tra
                             participantId, conferenceId);
                         await _videoPlatformService.TransferParticipantAsync(conferenceId,
                             transferRequest.ParticipantId,
-                            RoomType.WaitingRoom, RoomType.HearingRoom);
+                            RoomType.WaitingRoom.ToString(), RoomType.HearingRoom.ToString());
```
Comment (rehmanab) [inline VideoApi/VideoApi/Controllers/ConferenceManagementController.cs:148]: null check

File: VideoApi/VideoApi/Controllers/ConferenceManagementController.cs [hunk 4]
```diff
@@ -145,14 +145,14 @@ public async Task<IActionResult> TransferParticipantAsync(Guid conferenceId, Tra
...
                         break;
                     case TransferType.Dismiss:
                         _logger.LogDebug("Attempting to transfer {Participant} out of hearing room in {Conference}",
                             participantId, conferenceId);
                         await _videoPlatformService.TransferParticipantAsync(conferenceId,
                             transferRequest.ParticipantId,
-                            RoomType.HearingRoom, RoomType.WaitingRoom);
+                            RoomType.HearingRoom.ToString(), RoomType.WaitingRoom.ToString());
```
Comment (rehmanab) [inline VideoApi/VideoApi/Controllers/ConferenceManagementController.cs:155]: null check

**Review by rehmanab — APPROVED** [review]

File: VideoApi/VideoApi/Controllers/ConsultationController.cs [hunk 5]
```diff
@@ -286,57 +257,42 @@ public async Task<IActionResult> StartConsultationRequestAsync(StartConsultation
         }
 
         /// <summary>
-        /// Leave a consultation.
+        /// Leave a consultation
         /// </summary>
         /// <param name="request"></param>
         /// <returns></returns>
-        [HttpPost("end")]
```
Comment (RuwanikaWeerakkody) [inline VideoApi/VideoApi/Controllers/ConsultationController.cs:293]: @levibutler  what is the reason for changing these? Start and End were supposed to be only for JOH consultations as they follow a different logic to participant consultations.  Implementation should have just removed/replaced the existing participant consultation logic and left the JOH endpoints as is.

File: VideoApi/VideoApi/Controllers/ConsultationController.cs [hunk 6]
```diff
@@ -241,14 +150,76 @@ public async Task<IActionResult> StartPrivateConsultationWithEndpointAsync(Endpo
...
+            catch (KinlyApiException ex)
+            {
+                _logger.LogError(ex,
+                    "Unable to create a consultation room for ConferenceId: {conferenceId}",
+                    request.ConferenceId);
+                return BadRequest("Consultation room creation failed");
+            }
         }
 
         [HttpPost("start")]
```
Comment (RuwanikaWeerakkody) [inline VideoApi/VideoApi/Controllers/ConsultationController.cs:220]: @levibutler  what is the reason for changing these? Start and End were supposed to be only for JOH consultations as they follow a different logic to participant consultations.  Implementation should have just removed/replaced the existing participant consultation logic and left the JOH endpoints as is.

File: VideoApi/VideoApi.Services/KinlyPlatformService.cs (hunk 1 above)
Comment (levibutler) [reply VideoApi/VideoApi.Services/KinlyPlatformService.cs:146]: this is enum, will never be null

File: VideoApi/VideoApi/Controllers/ConferenceManagementController.cs (hunk 3 above)
Comment (levibutler) [reply VideoApi/VideoApi/Controllers/ConferenceManagementController.cs:148]: enum, will never be null

File: VideoApi/VideoApi.UnitTests/Validation/AdminConsultationRequestValidationTests.cs [hunk 7]
```diff
@@ -10,14 +10,14 @@
 
 namespace VideoApi.UnitTests.Validation
 {
-    public class AdminConsultationRequestValidationTests
+    public class ConsultationRequestResponseValidationTests
```
Comment (levibutler) [inline VideoApi/VideoApi.UnitTests/Validation/AdminConsultationRequestValidationTests.cs:13]: Update file name to match

File: VideoApi/VideoApi.IntegrationTests/Features/Consultations/LeaveConsultation.feature [hunk 8]
```diff
@@ -8,4 +8,4 @@ Scenario: Leave Consultation for a Judge Joh
 	And the judge joh is in the consultation room
 	And I have a valid leave consultation request
 	When I send the request to the endpoint
-	Then the response should have the status Ok and success status True
+	Then the response should have the status NoContent and success status True
```
Comment (levibutler) [inline VideoApi/VideoApi.IntegrationTests/Features/Consultations/LeaveConsultation.feature:11]: Ok

File: VideoApi/VideoApi.UnitTests/Controllers/Consultation/LeaveConsultationTests.cs [hunk 9]
```diff
@@ -53,16 +59,17 @@ public async Task should_return_judge_joh_to_waiting_room_for_valid_conference_a
...
                     (leaveConsultationRequest.ConferenceId, leaveConsultationRequest.ParticipantId, fromRoom, toRoom),
+                Times.Never);
+            VideoPlatformServiceMock.Verify(v => v.TransferParticipantAsync(conferenceId, participantId, fromRoom, toRoom),
                 Times.Once);
-            VideoPlatformServiceMock.VerifyNoOtherCalls();
         }
 
         [Test]
-        public async Task Should_Return_BadRequest_When_Participant_Cannot_Be_Found()
+        public async Task Should_Return_NoContent_When_Participant_Cannot_Be_Found_In_Consultation_Room()
```
Comment (levibutler) [inline VideoApi/VideoApi.UnitTests/Controllers/Consultation/LeaveConsultationTests.cs:68]: change

File: VideoApi/VideoApi.UnitTests/Controllers/Consultation/LeaveConsultationTests.cs [hunk 10]
```diff
@@ -53,16 +59,17 @@ public async Task should_return_judge_joh_to_waiting_room_for_valid_conference_a
                 (It.IsAny<GetConferenceByIdQuery>()), Times.Once);
             ConsultationService.Verify(v => v.LeaveConsultationAsync
                     (leaveConsultationRequest.ConferenceId, leaveConsultationRequest.ParticipantId, fromRoom, toRoom),
+                Times.Never);
+            VideoPlatformServiceMock.Verify(v => v.TransferParticipantAsync(conferenceId, participantId, fromRoom, toRoom),
```
Comment (levibutler) [inline VideoApi/VideoApi.UnitTests/Controllers/Consultation/LeaveConsultationTests.cs:63]: remove, use conference service

File: VideoApi/VideoApi.UnitTests/Controllers/Consultation/RespondToAdminConsultationRequestTests.cs [hunk 11]
```diff
@@ -1,39 +1,37 @@
...
+using System;
+using System.Threading.Tasks;
 using VideoApi.Contract.Requests;
-using VideoApi.Extensions;
 
 namespace VideoApi.UnitTests.Controllers.Consultation
 {
     [TestFixture]
-    public class RespondToAdminConsultationRequestTests : ConsultationControllerTestBase
+    public class RespondToConsultationRequestResponseTests : ConsultationControllerTestBase
```
Comment (levibutler) [inline VideoApi/VideoApi.UnitTests/Controllers/Consultation/RespondToAdminConsultationRequestTests.cs:12]: update file name to reflect

**Review by levibutler — COMMENTED** [review]
I made it out the other side! great team effort here 🥇  thanks to everyone who has contributed 👍 
some comments

File: VideoApi/VideoApi.UnitTests/Controllers/Conference/BookNewConferenceTests.cs [hunk 12]
```diff
@@ -70,18 +70,6 @@ public async Task Should_verify_double_booking_for_given_conference_id()
             VideoPlatformServiceMock.Verify(v => v.GetVirtualCourtRoomAsync(It.IsAny<Guid>()), Times.Once);
             CommandHandlerMock.Verify(c => c.Handle(It.IsAny<UpdateMeetingRoomCommand>()), Times.Never);
         }
-        
-        [Test]
```
[the remark this answers is not in the document]
Comment (levibutler) [reply VideoApi/VideoApi.UnitTests/Controllers/Conference/BookNewConferenceTests.cs:74]: readded

File: VideoApi/VideoApi.Services/Contracts/IVideoPlatformService.cs [hunk 13]
```diff
@@ -15,27 +15,15 @@ public interface IVideoPlatformService
...
-        
-        /// <summary>
-        /// Returns participants in a given room to waiting room
+        /// Transfers all participants and endpoints to the waiting room (triggering kinly to close the room)
         /// </summary>
         /// <param name="conference"></param>
         /// <param name="consultationRoom"></param>
         /// <returns></returns>
-        Task StopPrivateConsultationAsync(Conference conference, RoomType consultationRoom);
+        Task StopConsultationAsync(Conference conference, string consultationRoom);
```
[the remark this answers is not in the document]
Comment (levibutler) [reply VideoApi/VideoApi.Services/Contracts/IVideoPlatformService.cs:26]: maybe used by the story @tremolo64 is working on?

**Review by RuwanikaWeerakkody — APPROVED** [review]

**Review by levibutler — COMMENTED** [review]
LGTM

**Review by valentinabennett — APPROVED** [review]

**Pull request merged**